libmuscle/python/libmuscle/manager/profile_database.py

```diff
@@ -10,27 +16,419 @@ class ProfileDatabase:
+        instance was active.
+
+        Note that sending messages in MUSCLE3 is partially done in the
+        background. Transfer times include encoding and queueing of any
+        sent messages as well as downloading received messages, but it
+        does not include the sending side of the transfer, as this is
+        done by a background thread in parallel with other work that is
+        recorded (usually waiting, sometimes computing or sending
+        another message).
+
+        Nevertheless, this should give you an idea of which instances
+        use the most resources. Keep in mind though that different
+        instances may use different numbers of cores, so a model that
+        doesn't spend much wallclock time may still spend many core
+        hours. Secondly, waiting does not necessarily leave a core idle
+        as MUSCLE3 is capable of detecting when models will not compute
+        at the same time and having them share cores.
+
+        See the profiling documentation page for an example.
+
+        Returns:
+            A list of instance names, a corresponding list of compute
+            times, a corresponding list of transfer times, and a
+            corresponding list of wait times.
+        """
+        cur = self._get_cursor()
+        cur.execute("BEGIN TRANSACTION")
+        cur.execute(
+                "SELECT instance, stop_time"
+                " FROM all_events"
+                " WHERE type = 'CONNECT'")
+        start_run = dict(cur.fetchall())
+
+        cur.execute(
+                "SELECT instance, start_time"
+                " FROM all_events"
+                " WHERE type = 'DEREGISTER'")
+        stop_run = dict(cur.fetchall())
+
+        cur.execute(
+                "SELECT instance, SUM(stop_time - start_time)"
+                " FROM all_events"
+                " WHERE type = 'SEND' OR type = 'RECEIVE'"
+                " GROUP BY instance")
+        comm = dict(cur.fetchall())
+
+        cur.execute(
+                "SELECT instance, SUM(stop_time - start_time)"
+                " FROM all_events"
+                " WHERE type = 'RECEIVE_WAIT'"
+                " GROUP BY instance")
+        wait = dict(cur.fetchall())
+
+        cur.execute("COMMIT")
+        cur.close()
+
+        instances = list(start_run.keys())
+        total_times = [(stop_run[i] - start_run[i]) * 1e-9 for i in instances]
+        comm_times = [
+                (
+                    (comm[i] if i in comm else 0) -
+                    (wait[i] if i in wait else 0)
+                ) * 1e-9
+                for i in instances]
+        wait_times = [(wait[i] if i in wait else 0) * 1e-9 for i in instances]
+        run_times = [
+                t - c - w
+                for t, c, w in zip(total_times, comm_times, wait_times)]
+
+        return instances, run_times, comm_times, wait_times
+
+    def resource_stats(self) -> Dict[str, Dict[str, float]]:
+        """Calculate per-core statistics.
```

Comment:
This assumes that a multi-core instance is distributing the workload evenly on all cores, right?

Reply:
Yes, essentially it takes those resources as occupied during that time, and whatever the model code is doing internally is its concern, and for some other software to measure. Although I'm sure people will eventually want some documentation on how to profile submodels within a MUSCLE3 simulation...